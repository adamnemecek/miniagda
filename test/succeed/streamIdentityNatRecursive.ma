
sized data SNat : Size -> Set
{
  zero : (i : Size) -> SNat ($ i);
  succ : (i : Size) -> SNat i -> SNat ($ i)
}

sized codata Stream : Size -> Set {
  cons : (i : Size) -> SNat # -> Stream i -> Stream ($ i)
}


-- a silly stream identity
-- this is refuted by Karl's overly restrictive admissibility test

fun sid : (i : Size) -> SNat i -> (j : Size) -> Stream j -> Stream j
{
  sid .($ i) (zero i)   j xs = xs ;
  sid .($ i) (succ i y) j xs = sid i y j xs
}
