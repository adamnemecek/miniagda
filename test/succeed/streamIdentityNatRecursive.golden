MiniAgda by Andreas Abel and Karl Mehltretter
--- opening "streamIdentityNatRecursive.ma" ---
--- scope checking ---
--- type checking ---
type  SNat : + Size -> Set
term  SNat.zero : .[s!ze : Size] -> .[i < s!ze] -> SNat s!ze
term  SNat.zero : .[i : Size] -> < SNat.zero i : SNat $i >
term  SNat.succ : .[s!ze : Size] -> .[i < s!ze] -> ^ SNat i -> SNat s!ze
term  SNat.succ : .[i : Size] -> ^(y1 : SNat i) -> < SNat.succ i y1 : SNat $i >
type  Stream : - Size -> Set
term  Stream.cons : .[i : Size] -> ^(y1 : SNat #) -> ^(y2 : Stream i) -> < Stream.cons i y1 y2 : Stream $i >
term  sid : .[i : Size] -> SNat i -> .[j : Size] -> Stream j -> Stream j
{ sid [.$i] (SNat.zero [i]) [j] xs = xs
; sid [.$i] (SNat.succ [i] y) [j] xs = sid [i] y [j] xs
}
--- evaluating ---
--- closing "streamIdentityNatRecursive.ma" ---
